src/dotnet/CarbonAware.LocationSources.Azure/src/AzureLocationSource.cs

```diff
@@ -0,0 +1,107 @@
+
+    private readonly ILogger<AzureLocationSource> _logger;
+
+    private IDictionary<string, NamedGeoposition> namedGeopositions;
+
+    private static readonly JsonSerializerOptions options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="AzureLocationSource"/> class.
+    /// </summary>
+    /// <param name="logger">The logger for the datasource</param>
+    public AzureLocationSource(ILogger<AzureLocationSource> logger)
+    {
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        namedGeopositions = new Dictionary<string, NamedGeoposition>();
+    }
+
+    public Location ToGeopositionLocation(Location location)
+    {
+        switch (location.LocationType)
+        {
+            case LocationType.Geoposition:
+            {
+                return location;
+            }
+            case LocationType.CloudProvider: 
+            {
+                if( location.CloudProvider != CloudProvider.Azure ) 
+                {
+                    throw new LocationConversionException($"Incorrect Cloud provider region. Expected Azure but found '{ location.CloudProvider }'");
+                }
+                
+                return getGeoPositionLocationOrThrow(location);
+            }
+        }
+        
+        throw new LocationConversionException($"Location '{ location.CloudProvider }' cannot be converted to Geoposition. ");
+    }
+
+    private Location getGeoPositionLocationOrThrow(Location location)
+    {
+        loadRegionsFromFileIfNotPresent();
+        
+        NamedGeoposition geopositionLocation = namedGeopositions[location.RegionName ?? ""];    
+        if(geopositionLocation == null) 
+        {
+            throw new ArgumentException($"Lat/long cannot be retrieved for region '{ location.RegionName }'");
+        }
+        if (_logger.IsEnabled(LogLevel.Debug))
+        {
+            _logger.LogDebug("Converted Azure Location named '{regionName}' to Geoposition Location at latitude '{latitude}'" 
+                                + "and logitude '{longitude}'.", location.RegionName, geopositionLocation.Latitude, geopositionLocation.Longitude);
+        }
+        return new Location 
+                {
+                    LocationType = LocationType.Geoposition,
+                    Latitude = Convert.ToDecimal(geopositionLocation.Latitude),
+                    Longitude = Convert.ToDecimal(geopositionLocation.Longitude)
+                };
+    }
+
+    protected virtual Dictionary<string, NamedGeoposition> LoadRegionsFromJson()
+    {
+        var data = ReadFromResource("CarbonAware.LocationSources.Azure.azure-regions.json");
+        List<NamedGeoposition> regionList = JsonSerializer.Deserialize<List<NamedGeoposition>>(data, options) ?? new List<NamedGeoposition>();
+        Dictionary<string, NamedGeoposition> namedGeopositions = new Dictionary<String, NamedGeoposition>();
```

Comment:
This is confusing using another instance from the attribute class (line 34). Either you instance it here or in the constructor, otherwise you're creating memory twice.